### PR TITLE
Add more password changing structures and calls

### PR DESCRIPTION
**Pull request by smashery**
Add structures for changing passwords. This is primarily to support https://github.com/rapid7/metasploit-framework/pull/19666

### CODE REVIEW

File: lib/ruby_smb/dcerpc/samr/samr_get_members_in_group_response.rb [hunk 1]
```diff
@@ -14,7 +14,7 @@ class PsamprGetMembersBuffer < SamprGetMembersBuffer
         extend Ndr::PointerClassPlugin
       end
 
-      # [2.1.5.8.3 SamrGetMembersInGroup (Opnum 25)](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-samr/a4adbf20-040f-4416-a960-e5b7917fdae7)
+      # [3.1.5.8.3 SamrGetMembersInGroup (Opnum 25)](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-samr/a4adbf20-040f-4416-a960-e5b7917fdae7)
```
Comment (smcintyre-r7) [inline lib/ruby_smb/dcerpc/samr/samr_get_members_in_group_response.rb:17]:
```suggestion
      # [3.1.5.8.3 SamrGetMembersInGroup (Opnum 25)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-samr/3ed5030d-88a3-42ca-a6e0-8c12aa2fdfbd)
```

I think the URL here is wrong too.

**Review by smcintyre-r7 — COMMENTED** [review]
Just a couple of minor comments. Over all this looks good so I'll start the testing with Metasploit.

File: lib/ruby_smb/dcerpc/samr.rb [hunk 2]
```diff
@@ -318,6 +320,58 @@ class PulongArray < Ndr::NdrConfArray
...
+            cipher.encrypt
+            cipher.key = key1
+          end
+
+          cipher2 = OpenSSL::Cipher.new('des-ecb').tap do |cipher|
+            cipher.encrypt
+            cipher.key = key2
+          end
+
+          cipher1.update(block1) + cipher2.update(block2)
```
[the remark this answers is not in the document]
Comment (smashery) [reply lib/ruby_smb/dcerpc/samr.rb:363]: Turns out this adds extra bytes on that we don't want.

File: lib/ruby_smb/dcerpc/samr.rb [hunk 3]
```diff
@@ -333,6 +387,28 @@ class RpcShortBlob < BinData::Record
...
+
+        def self.encrypt_password(password, old_password_nt)
+          # see: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-samr/5fe3c4c4-e71b-440d-b2fd-8448bfaf6e04
+          password = password.encode('UTF-16LE').force_encoding('ASCII-8BIT')
+          buffer = password.rjust(512, "\x00") + [ password.length ].pack('V')
+          cipher = OpenSSL::Cipher.new('RC4').tap do |cipher|
+            cipher.encrypt
+            cipher.key = old_password_nt
+          end
+          cipher.update(buffer)
```
[the remark this answers is not in the document]
Comment (smashery) [reply lib/ruby_smb/dcerpc/samr.rb:403]: As above - adding this breaks the code.

**Pull request merged**